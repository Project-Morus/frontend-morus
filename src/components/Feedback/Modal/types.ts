import { ReactNode } from "react";
import { IBaseProps } from "../../../types";
import { IconNamesTypes } from "../../DataDisplay/Icon/types";

export interface IBackdropProps extends IBaseProps {
  children: ReactNode;
  open: boolean;
}

export interface IModalProps extends IFormModalProps {
  open: boolean;
  variant?: VariantsTypes;
  children?: ReactNode;
  closeModal: () => void;
  onConfirmModal: () => void;
}

export interface IModalVariantsProps extends IBaseProps {
  children?: ReactNode;
  onConfirmModal: () => void;
  confirmButtonName?: string;
  closeModal: () => void;
  buttonLabel?: string;
  isLoading?: boolean;
}

export interface IFormModalProps extends IModalVariantsProps {
  iconName?: IconNamesTypes;
  modalTitle?: string;
}

export type VariantsTypes = "confirmation" | "warning" | "form";
