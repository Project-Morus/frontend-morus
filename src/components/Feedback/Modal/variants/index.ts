export * from './Confirmation';
export * from './Form';
export * from './Warning';