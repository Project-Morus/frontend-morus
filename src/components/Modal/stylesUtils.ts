import styled from "styled-components";

export const WIDTH_CONFIRMATION_MODAL = '620px'

export const BASE_STYLE = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: ${({ theme }) => theme.spacing[7]};
  font-size: ${({ theme }) => theme.typography.TEXT_SIZES.large};
  border-radius: ${({ theme }) => theme.border.radius.large};
  background-color: ${({ theme }) => theme.colors.white[100]};
  margin: auto;
`;

export const BackdropSC = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.63);
  display: flex;
  justify-content: center;
  align-items: center;
`;
