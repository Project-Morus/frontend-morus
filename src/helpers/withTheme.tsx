import { ComponentType, JSX } from 'react';
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "../styles/themes/default";

type ComponentPropsType<T> = T & JSX.IntrinsicAttributes

export function withTheme<T>(WrapperComponent: ComponentType<ComponentPropsType<T>>) {
  return (props: ComponentPropsType<T>) => {
    return (
      <ThemeProvider theme={defaultTheme}>
        <WrapperComponent {...props} />
      </ThemeProvider>
    );
  };
}

