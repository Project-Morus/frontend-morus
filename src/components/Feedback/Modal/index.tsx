import { ReactNode } from "react";
import { IModalProps, VariantsTypes } from "./types";
import Backdrop from "./parts/Backdrop";
import { ConfirmationModal, WarningModal, FormModal } from "./variants";

const Modal = ({
  variant = "confirmation",
  closeModal,
  children,
  onConfirmModal,
  iconName,
  modalTitle,
  confirmButtonName,
  buttonLabel,
  open,
  isLoading,
}: IModalProps) => {
  const variantType: Record<VariantsTypes, ReactNode> = {
    confirmation: (
      <ConfirmationModal
        role="modal"
        onConfirmModal={onConfirmModal}
        confirmButtonName={confirmButtonName}
        buttonLabel={buttonLabel}
        closeModal={closeModal}
        isLoading={isLoading}
      >
        {children}
      </ConfirmationModal>
    ),
    warning: (
      <WarningModal
        role="modal"
        onConfirmModal={onConfirmModal}
        confirmButtonName={confirmButtonName}
        buttonLabel={buttonLabel}
        closeModal={closeModal}
        isLoading={isLoading}
      >
        {children}
      </WarningModal>
    ),
    form: (
      <FormModal
        role="modal"
        onConfirmModal={onConfirmModal}
        closeModal={closeModal}
        iconName={iconName}
        modalTitle={modalTitle}
        confirmButtonName={confirmButtonName}
        buttonLabel={buttonLabel}
        isLoading={isLoading}
      >
        {children}
      </FormModal>
    ),
  };

  return (
    <Backdrop role="backdrop" open={open}>
      {variantType[variant]}
    </Backdrop>
  );
};

export default Modal;
