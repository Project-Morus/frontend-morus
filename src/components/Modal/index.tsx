import { ReactNode } from "react";
import { IModalProps, VariantsTypes } from "./types";
import Backdrop from "./parts/Backdrop";
import { ConfirmationModal, WarningModal, FormModal } from "./variants";


const Modal = ({
  variant = "confirmation",
  closeModal,
  children,
  onConfirmModal,
  iconName,
  modalTitle,
  confirmButtonName,
  label
}: IModalProps) => {
  const variantType: Record<VariantsTypes, ReactNode> = {
    confirmation: (
      <ConfirmationModal
        onConfirmModal={onConfirmModal}
        confirmButtonName={confirmButtonName}
        label={label}>
        {children}
      </ConfirmationModal>
    ),
    warning: (
      <WarningModal
        onConfirmModal={onConfirmModal}
        confirmButtonName={confirmButtonName}
        label={label}>
        {children}
      </WarningModal>
    ),
    form: (
      <FormModal
        onConfirmModal={onConfirmModal}
        iconName={iconName}
        modalTitle={modalTitle}
        confirmButtonName={confirmButtonName}
        label={label}
      >
        {children}
      </FormModal>
    ),
  };

  return <Backdrop closeModal={closeModal}>{variantType[variant]}</Backdrop>;
};

export default Modal;
