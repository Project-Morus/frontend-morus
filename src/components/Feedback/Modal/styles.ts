import styled from "styled-components";
import { BASE_STYLE, WIDTH_CONFIRMATION_MODAL } from "./stylesUtils";

export const ButtonsContainerSC = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: ${({ theme }) => theme.spacing[4]};

  margin-top: auto;
`;

export const ChildrenContainerSC = styled.div`
  width: 100%;
  margin: ${({ theme }) => theme.spacing[8]} 0;

  color: ${({ theme }) => theme.colors.grey[300]};
  font-size: ${({ theme }) => theme.typography.TEXT_SIZES.medium};
`;

export const TitleSC = styled.h2`
  font-size: ${({ theme }) => theme.typography.HEADERS_SIZES.h2};
  color: ${({ theme }) => theme.colors.grey[400]};
`;

export const HeaderSC = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  gap: ${({ theme }) => theme.spacing[8]};
`;

export const ConfirmationModalSC = styled(BASE_STYLE)`
  background-color: white;
  width: 100%;
  max-width: ${WIDTH_CONFIRMATION_MODAL};
  overflow: auto;
  text-align: center;
`;

export const FormModalSC = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: ${({ theme }) => theme.spacing[7]};
  font-size: ${({ theme }) => theme.typography.TEXT_SIZES.large};
  border-radius: ${({ theme }) => theme.border.radius.large};
  background-color: ${({ theme }) => theme.colors.white[100]};
  margin: auto;
`;
