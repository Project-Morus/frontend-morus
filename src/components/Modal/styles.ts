import styled from "styled-components";
import { BASE_STYLE } from "./stylesUtils";

export const ButtonsConainterSC = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const ChildrenContainerSC = styled.div`
  margin: ${({ theme }) => theme.spacing[8]} 0;
  text-align: center;
`;

export const ConfirmationModalSC = styled(BASE_STYLE)`
  background-color: white;
  width: 620px;
  overflow: auto;
`;

export const IconAndTitleModalSC = styled(BASE_STYLE)`
  background-color: white;
`;
