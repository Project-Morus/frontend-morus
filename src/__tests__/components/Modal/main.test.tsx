import { render, screen } from "@testing-library/react";
import Modal from "../../../components/Modal";
import { withTheme } from "../../../helpers/withTheme";

const ModalWithTheme = withTheme(Modal);
const modalElement = "modal";
const modalChildrenText = "Teste";

const closeModalFn = jest.fn();
const confirmModalFn = jest.fn();

describe("Modal Component", () => {
  let container: HTMLElement;

  beforeEach(() => {
    container = document.createElement("div");
    container.id = "portal";
    document.body.appendChild(container);
  });

  afterEach(() => {
    if (container) {
      container.remove();
    }
  });
  test("Verify if component is rendering", () => {
    render(
      <ModalWithTheme open closeModal={closeModalFn} onConfirmModal={confirmModalFn} role="modal">
        <h1>{modalChildrenText}</h1>
      </ModalWithTheme>,
      { container }
    );

    expect(screen.getByRole(modalElement)).toBeVisible();
    expect(screen.getByText(modalChildrenText)).toBeInTheDocument();
  });

  test("Snapshot Modal", () => {
    const { asFragment } = render(
      <ModalWithTheme open closeModal={closeModalFn} onConfirmModal={confirmModalFn} role="modal">
        <h1>{modalChildrenText}</h1>
      </ModalWithTheme>,
      { container }
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("Snapshot Modal with variant warning", () => {
    const { asFragment } = render(
      <ModalWithTheme open closeModal={closeModalFn} onConfirmModal={confirmModalFn} role="modal" variant="warning">
        <h1>{modalChildrenText}</h1>
      </ModalWithTheme>,
      { container }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Snapshot Modal with variant form", () => {
    const { asFragment } = render(
      <ModalWithTheme
        open
        closeModal={() => {}}
        onConfirmModal={() => {}}
        role="modal"
        variant="form"
        iconName="ph-user"
        modalTitle="Modal Content"
      >
        <h1>Modal Content</h1>
      </ModalWithTheme>,
      { container }
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
