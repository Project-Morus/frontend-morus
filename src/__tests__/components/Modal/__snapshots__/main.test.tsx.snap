// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component Snapshot Modal 1`] = `
<DocumentFragment>
  .c6 {
  width: 100%;
  height: 50px;
  font-weight: 700;
  border: none;
  border-radius: 5px;
  background: none;
  cursor: pointer;
  transition: all 0.2s;
  max-width: 200px;
}

.c6:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c7 {
  border: 2px solid #0353A4;
  color: #0353A4;
}

.c7:not(:disabled):hover {
  border-color: #034C95;
  background-color: #0353A4;
  color: #FFF;
}

.c8 {
  border: 2px solid #3F9D50;
  background-color: #3F9D50;
  color: #FFF;
}

.c8:not(:disabled):hover {
  border-color: #398F49;
  background-color: #398F49;
  color: #FFF;
}

.c1 {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1.5rem;
  font-size: 1.25rem;
  border-radius: 10px;
  background-color: #F8F8F8;
  margin: auto;
}

.c0 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.43);
  display: flex;
  justify-content: center;
  align-items: center;
}

.c5 {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c4 {
  margin: 1.75rem 0;
  text-align: center;
  color: #6B6B6B;
  font-size: 1rem;
}

.c3 {
  font-size: 1.8125rem;
  color: #4E4E4E;
}

.c2 {
  background-color: white;
  width: 100%;
  max-width: 620px;
  overflow: auto;
}

<div
    class="c0"
    role="backdrop"
  >
    <div
      class="c1 c2"
      role="modal"
    >
      <h2
        class="c3"
      >
        Atenção!
      </h2>
      <div
        class="c4"
      >
        <h1>
          Teste
        </h1>
      </div>
      <div
        class="c5"
      >
        <button
          class="c6 c7"
        >
          <span>
            Cancelar
          </span>
        </button>
        <button
          class="c6 c8"
        >
          <span>
            Confirmar
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal Component Snapshot Modal with variant form 1`] = `
<DocumentFragment>
  .c9 {
  width: 100%;
  height: 50px;
  font-weight: 700;
  border: none;
  border-radius: 5px;
  background: none;
  cursor: pointer;
  transition: all 0.2s;
  max-width: 200px;
}

.c9:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c11 {
  background-color: #0353A4;
  color: #FFF;
}

.c11:not(:disabled):hover {
  background-color: #034C95;
}

.c10 {
  border: 2px solid #0353A4;
  color: #0353A4;
}

.c10:not(:disabled):hover {
  border-color: #034C95;
  background-color: #0353A4;
  color: #FFF;
}

.c5 {
  color: #0353A4;
}

.c6 {
  display: flex;
  align-items: center;
  justify-content: center;
  max-width: 100%;
  width: fit-content;
  width: -moz-fit-content;
  height: fit-content;
  height: -moz-fit-content;
}

.c4 {
  width: 70px;
  height: 70px;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: #E6EEF6;
  box-shadow: 0px 4px 4px 0px rgba(0, 0, 0, 0.05);
  border-radius: 5px;
}

.c1 {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1.5rem;
  font-size: 1.25rem;
  border-radius: 10px;
  background-color: #F8F8F8;
  margin: auto;
}

.c0 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.43);
  display: flex;
  justify-content: center;
  align-items: center;
}

.c8 {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c7 {
  margin: 1.75rem 0;
  text-align: center;
  color: #6B6B6B;
  font-size: 1rem;
}

.c3 {
  width: 100%;
  display: flex;
  align-items: center;
  gap: 1.75rem;
}

.c2 {
  background-color: white;
}

<div
    class="c0"
    role="backdrop"
  >
    <div
      class="c1 c2"
      role="modal"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              class="c6"
              data-testid="ph-user"
            >
              <svg
                fill="currentColor"
                height="47"
                viewBox="0 0 256 256"
                width="47"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M230.92,212c-15.23-26.33-38.7-45.21-66.09-54.16a72,72,0,1,0-73.66,0C63.78,166.78,40.31,185.66,25.08,212a8,8,0,1,0,13.85,8c18.84-32.56,52.14-52,89.07-52s70.23,19.44,89.07,52a8,8,0,1,0,13.85-8ZM72,96a56,56,0,1,1,56,56A56.06,56.06,0,0,1,72,96Z"
                />
              </svg>
            </div>
          </div>
        </div>
        <h1>
          Modal Content
        </h1>
      </div>
      <div
        class="c7"
      >
        <h1>
          Modal Content
        </h1>
      </div>
      <div
        class="c8"
      >
        <button
          class="c9 c10"
        >
          <span>
            Cancelar
          </span>
        </button>
        <button
          class="c9 c11"
        >
          <span>
            Confirmar
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal Component Snapshot Modal with variant warning 1`] = `
<DocumentFragment>
  .c5 {
  width: 100%;
  height: 50px;
  font-weight: 700;
  border: none;
  border-radius: 5px;
  background: none;
  cursor: pointer;
  transition: all 0.2s;
  max-width: 200px;
}

.c5:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c6 {
  border: 2px solid #0353A4;
  color: #0353A4;
}

.c6:not(:disabled):hover {
  border-color: #034C95;
  background-color: #0353A4;
  color: #FFF;
}

.c7 {
  border: 2px solid #CE323B;
  background-color: #CE323B;
  color: #FFF;
}

.c7:not(:disabled):hover {
  border-color: #BB2E36;
  background-color: #BB2E36;
  color: #FFF;
}

.c1 {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1.5rem;
  font-size: 1.25rem;
  border-radius: 10px;
  background-color: #F8F8F8;
  margin: auto;
}

.c0 {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.43);
  display: flex;
  justify-content: center;
  align-items: center;
}

.c4 {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.c3 {
  margin: 1.75rem 0;
  text-align: center;
  color: #6B6B6B;
  font-size: 1rem;
}

.c2 {
  background-color: white;
  width: 100%;
  max-width: 620px;
  overflow: auto;
}

<div
    class="c0"
    role="backdrop"
  >
    <div
      class="c1 c2"
      role="modal"
    >
      <h2>
        Atenção!
      </h2>
      <div
        class="c3"
      >
        <h1>
          Teste
        </h1>
      </div>
      <div
        class="c4"
      >
        <button
          class="c5 c6"
        >
          <span>
            Cancelar
          </span>
        </button>
        <button
          class="c5 c7"
        >
          <span>
            Excluir
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
